src/Halogen/Components/Select/Utils/NonEmpty.purs

```diff
@@ -0,0 +1,40 @@
+module Halogen.Components.Select.Utils.NonEmpty where
+
+import Prelude
+
+import Control.Alternative (Alternative)
+import Data.NonEmpty (NonEmpty(), (:|), oneOf)
+import Data.Maybe as M
+import Data.Maybe.Unsafe as Mu
+import Data.Array as A
+
+-- uncons empty = Nothing
+-- uncons (pure x <|> tail) = Just { head: x, tail: tail }
+class (Alternative f) <= Uncons f where
+  uncons :: forall a. f a -> M.Maybe {head :: a, tail :: f a}
+
+instance arrayUncons :: Uncons Array where
+  uncons = A.uncons
+
+liftNonEmpty
+  :: forall f a b
+   . (Uncons f)
+  => (f a -> f b)
+  -> NonEmpty f a
+  -> NonEmpty f b
+liftNonEmpty fab nefa =
+  let
+    fa :: f a
+    fa = oneOf nefa
+
+    fb :: f b
+    fb = fab fa
+
+    mbUnconsed :: M.Maybe {head :: b, tail :: f b}
+    mbUnconsed = uncons fb
+
+    -- Safe, `isJust (uncons (pure x <|> tail)) <-- oneOf === (pure x <|> tail)`
+    unconsed :: {head :: b, tail :: f b}
+    unconsed = Mu.fromJust mbUnconsed
```

Comment:
Maybe instead of using `NonEmpty` directly we could use some kind of lazy stream that is constructed from a `NonEmpty`, that is equivalent to a circular linked list and allows `takeLeft`/`takeRight` and does the rotation internally? Perhaps that's overly "clever" though without needing to be.